package/gluon-mesh-babel/src/babelparse.c

```diff
@@ -0,0 +1,200 @@
+
+	if (n != 7)
+		goto end;
+
+	struct in6_addr address;
+	if (inet_pton(AF_INET6, address_str, &address) != 1)
+	{
+		fprintf(stderr, "babeld-parser error: could not convert babel data to ipv6 address: %s\n", address_str);
+		goto end;
+	}
+	ret.action = action;
+	ret.address_str = address_str;
+	ret.ifname = ifname;
+	ret.reach = reach;
+	ret.rxcost = rxcost;
+	ret.txcost = txcost;
+	ret.cost = cost;
+
+end:
+	free(action);
+	free(address_str);
+	free(ifname);
+	return(ret);
+}
+
+struct babelroute babelparse_get_route(char *line) {
+	struct babelroute ret;
+	char *action = NULL;
+	char *route = NULL;
+	char *prefix = NULL;
+	char *from = NULL;
+	char *id = NULL;
+	int metric;
+	int refmetric ;
+	char *via = NULL;
+	char *ifname = NULL;
+
+	int n = sscanf(line, "%s route %s prefix %s from %s installed yes id %s metric %d refmetric %d via %s if %s",
+			&action, &route, &prefix, &from, &id, &metric, &refmetric, &via, &ifname );
+
+	if (n != 9)
+		goto end;
+
+	struct in6_addr in6_via;
+	if (inet_pton(AF_INET6, via, &in6_via) != 1)
+	{
+		fprintf(stderr, "babeld-parser error: could not convert babel data to ipv6 address: %s\n", via);
+		goto end;
+	}
+
+	ret.action = action;
+	ret.route = route;
+	ret.prefix = prefix;
+	ret.from = from;
+	ret.id = id;
+	ret.metric = metric;
+	ret.refmetric = refmetric;
+	ret.via = via;
+	ret.ifname = ifname;
+	ret.in6_via = in6_via;
+
+end:
+	free(action);
+	free(route);
+	free(prefix);
+	free(from);
+	free(id);
+	free(via);
+	free(ifname);
+	return(ret);
+}
+
+static void input_pump(int fd, struct json_object* obj, void (*lineprocessor)(char*, struct json_object*)) {
+	char inbuf[SOCKET_INPUT_BUFFER_SIZE];
+	size_t inbuf_used = 0;
+	size_t inbuf_remain = 0;
+	char *line_start = &inbuf[0] ;
+	ssize_t rv=0;
+	char endstring[3];
+
+	snprintf(endstring, 3, "ok");
+	memset(inbuf, '\0', sizeof(inbuf));
+
+	while  (strncmp(line_start, endstring, strlen(endstring))) {
+		inbuf_remain = sizeof(inbuf) - inbuf_used;
+		if (inbuf_remain == 0) {
+			fprintf(stderr, "Line exceeded buffer length. recompile respondd with larger SOCKET_INPUT_BUFFER_SIZE\n");
+			return;
+		}
+		rv=0;
+		while (rv <=0 ) {
+			rv = recv(fd, (void*)&inbuf[inbuf_used], inbuf_remain, MSG_DONTWAIT);
+			if (rv == 0) {
+				// fprintf(stderr, "Connection closed.\n"); // this may or may not be an error => donot print a message
+				return;
+			}
+			else if (rv < 0 && errno == EAGAIN) {
+				usleep(1000);
+			}
+			else if (rv < 0) {
+				perror("Connection error");
+			}
+		}
+		inbuf_used += rv;
+
+		//scan for newlines, have whole lines processed by lineprocessor
+		line_start = inbuf;
+		char *line_end;
+		while ( (line_end = (char*)memchr((void*)line_start, '\n', inbuf_used - (line_start -  inbuf))))
+		{
+			*line_end = 0;
+			if (lineprocessor)
+				lineprocessor(line_start, obj);
+			if (!strncmp(line_start, endstring, strlen(endstring)))
+				return;
+			line_start = line_end + 1;
```

Comment:
I think this could lead to a negative `n` passed to `memchr()`? Couldn't this blow up at the cast, since the parameter `n` is a `size_t` (aka. non negative)? Are you aware of this? Not really sure, whether that's in fact a problem.